### PR TITLE
PhoneInfoga#995

### LINKED ISSUE #995: [Analyzer] phoneinfoga
## Name
PhoneInfoga is one of the most advanced tools to scan international phone numbers

## Link
https://github.com/sundowndev/phoneinfoga

## Type of analyzer
observable, generic.

### PR DESCRIPTION
**Pull request by g4ze**
closes #995 
# Description
Add phoneinfoga_scan analyzer

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [x] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [x] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [x] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [x] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [x] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.
Screenshot from 2024-01-31 21-10-25

raw json:
{"result":{"valid":true,"number":"33679368229","local_format":"0679368229","international_format":"+33679368229","country_prefix":"+33","country_code":"FR","country_name":"France","location":"","carrier":"Orange France SA","line_type":"mobile"}}

### CODE REVIEW

**Comment by g4ze** [comment]
I still need to update docs and make migrations

**Comment by g4ze** [comment]
Docs done

**Comment by g4ze** [comment]
Hello! @mlodic , this took me sometime to get done. The test is failing here. It pertains to the migration file. I used the dumpplugin comand to create the file. I've crossed checked it with the way previous migrations are written. Still can't seem to find out why this is happening.

File: integrations/phoneinfoga/compose.yml
Comment (mlodic) [inline integrations/phoneinfoga/compose.yml:15]: reference to this?

File: integrations/phoneinfoga/compose-tests.yml
Comment (mlodic) [inline integrations/phoneinfoga/compose-tests.yml:13]: copy/pastes are not required because the `compose-tests.yml` is an override of the `compose.yml`. Please use other docker analyzers as a reference

File: integrations/phoneinfoga/compose.yml
Comment (mlodic) [inline integrations/phoneinfoga/compose.yml:6]: please deploy only the REST API, as stated in [the doc](https://sundowndev.github.io/phoneinfoga/getting-started/usage/). The GUI is not required and it is heavy

File: docs/source/Advanced-Usage.md
Comment (mlodic) [inline docs/source/Advanced-Usage.md:136]: can you provide an example for other scanners? to see if they work as intended

File: api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py
Comment (mlodic) [inline api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py:39]: please provide link to the official doc so the people can find it and choose the scanner type easier

File: api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py
Comment (mlodic) [inline api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py:52]: I guess you wanted to put `local`, right?

**Review by mlodic — CHANGES_REQUESTED** [review]

File: docs/source/Advanced-Usage.md
Comment (g4ze) [reply docs/source/Advanced-Usage.md:136]: do you mean that i should provide screenshot for other scanners?

File: api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py
Comment (g4ze) [reply api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py:52]: No, this is what i intended:
image

File: docs/source/Advanced-Usage.md
Comment (g4ze) [reply docs/source/Advanced-Usage.md:136]: this?
image

File: integrations/phoneinfoga/compose.yml
Comment (g4ze) [reply integrations/phoneinfoga/compose.yml:15]: this should run the GIN framework used in release mode. It works fine even without it.

File: integrations/phoneinfoga/compose-tests.yml
Comment (g4ze) [reply integrations/phoneinfoga/compose-tests.yml:13]: so should this file even contain anything as we are using an external service; would it make sense to have an additional dockerfile that clones the repo and builds it for the `compose-tests.yml`?

File: docs/source/Advanced-Usage.md
Comment (mlodic) [reply docs/source/Advanced-Usage.md:136]: ty, cool

File: api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py
Comment (mlodic) [reply api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py:52]: that's good, my bad

File: integrations/phoneinfoga/compose.yml
Comment (mlodic) [inline integrations/phoneinfoga/compose.yml:1]: use version 3.8 if you can, like we do in all other docker compose files in the project

File: integrations/phoneinfoga/compose-tests.yml
Comment (mlodic) [reply integrations/phoneinfoga/compose-tests.yml]: you are right, we could basically leave this almost empty.

Comment (g4ze) [reply integrations/phoneinfoga/compose-tests.yml]: should I just mention `version: 3.8` in the test yml and rest empty?

**Comment by mlodic** [comment]
the migration error is here:
```
  File "/opt/deploy/intel_owl/api_app/analyzers_manager/migrations/0064_analyzer_config_phoneinfoga.py", line 118, in migrate
    _create_object(PluginConfig, value)

```
I think that there is some problem with the IDs
````django.db.utils.IntegrityError: duplicate key value violates unique constraint "api_app_parameter_name_python_module_id_85f61e18_uniq"
```

I think that could be the error, I asking to @0ssigeno  a second opinion

**Comment by 0ssigeno** [comment]
Checking

**Comment by g4ze** [comment]
thankyou very much, I'll look into it.

**Comment by g4ze** [comment]
the tests are now successful.
`Error: Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3.1.0/dist/codecov' failed with exit code 255` is this on my end?
kindly review.

**Comment by g4ze** [comment]
apart from this I'll update the migrations as a new analyzer Zippy_scan was added.

**Comment by g4ze** [comment]
uhoh, i did it again; overlooked a checkpoint. Made required migrations(agn)

**Comment by g4ze** [comment]
it worked lol. @mlodic  any other changes?

File: api_app/playbooks_manager/migrations/0027_add_phoneinfoga_scan_to_free_to_use_analyzers.py
Comment (mlodic) [inline api_app/playbooks_manager/migrations/0027_add_phoneinfoga_scan_to_free_to_use_analyzers.py:34]: This is not required because this analyzer is "optional" and it requires to run an additional container.
To avoid to have it providing errors in the base installation, that is the most commonly used, I would not add it in the default playbook. Please remove.

File: docker/env_file_integrations_template
Comment (mlodic) [inline docker/env_file_integrations_template:8]: isn't there a way to leverage classic Analyzer Secrets for this API key? In that way, every user can configure it and use its own API key.
I don't like to have it configured here because this API key would be shared among all the users. This could be an unwanted behavior.

File: api_app/playbooks_manager/migrations/0027_add_phoneinfoga_scan_to_free_to_use_analyzers.py
Comment (g4ze) [reply api_app/playbooks_manager/migrations/0027_add_phoneinfoga_scan_to_free_to_use_analyzers.py:34]: noted. Shouldn't we also mention that such type of analyzers which are free but `docker-based`/`optional ` can't go in default playbook in the `docs` or the `PR checklist`; to avoid such implementations in future.

Comment (mlodic) [reply api_app/playbooks_manager/migrations/0027_add_phoneinfoga_scan_to_free_to_use_analyzers.py:34]: sure it would make sense to explicitly say it in the PR template

Comment (g4ze) [reply api_app/playbooks_manager/migrations/0027_add_phoneinfoga_scan_to_free_to_use_analyzers.py:34]: just one more concern, about another docker based analyzer, cyberchef which is in the default playbook
image

File: docker/env_file_integrations_template
Comment (g4ze) [reply docker/env_file_integrations_template:8]: hi there! I took a look at their code base. The problem at hand involves the extraction of all the API keys from the from the `integration .env` file instead of directly using the key as authorization header when making calls. This is hard coded in their code base. 
One work around could be making a new container everytime a new user selects the analyzer. This would have some infra implications but would allow us to have unique containers of the analyzer with its own env file for each individual user. (not a good approach tho)
I have raised an (issue)[https://github.com/sundowndev/phoneinfoga/issues/1393] for further enquiry.
In the meantime we could update the analyzer configs to take API keys from the user in the interface and implement that logic whenever they solve our problem, till then we might have to leverage `integration env` only.

Comment (mlodic) [reply docker/env_file_integrations_template:8]: thank you for the efforts. I would like to avoid to create complex things for this problem, I like to keep things simple if possible.
I suggest waiting for an answer from them. In case they do not answer within a week, we consider the problem unsolved and so we'll merge your actual PR even if there is that problem. In that case, we would need to specify more information about it in the docs, just to be explicit about this problem with the users.

Comment (g4ze) [reply docker/env_file_integrations_template:8]: for sure! In the meantime can you guide me to another issue i can take up?

Comment (mlodic) [reply docker/env_file_integrations_template:8]: this is a sightly different analyzer: https://github.com/intelowlproject/IntelOwl/issues/1103.
This is a really important source for this project

**Comment by mlodic** [comment]
Just to keep a reference, they answered [here](https://github.com/sundowndev/phoneinfoga/issues/1393)https://github.com/https://github.com/sundowndev/phoneinfoga/issues/1393)

We'll wait them to solve the issue regarding API keys than we'll finish this PR

**Comment by g4ze** [comment]
image
works using new version of phoneinfoga

**Comment by g4ze** [comment]
@mlodic kindly review.

**Review by mlodic — APPROVED** [review]

**Comment by mlodic** [comment]
let's wait for the tests. If they pass we can merge the PR. Ty!

**Pull request merged**
